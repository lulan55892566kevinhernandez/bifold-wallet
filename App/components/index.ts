import Button from './buttons/Button'
import CheckBoxRow from './inputs/CheckBoxRow'
import QRScanner from './inputs/QRScanner'
import TextInput from './inputs/TextInput'
import ContactListItem from './listItems/ContactListItem'
import CredentialListItem from './listItems/CredentialListItem'
import NotificationCredentialListItem from './listItems/NotificationCredentialListItem'
import NotificationProofListItem from './listItems/NotificationProofListItem'
import Failure from './modals/Failure'
import Message from './modals/Message'
import Pending from './modals/Pending'
import Success from './modals/Success'
import Label from './texts/Label'
import Text from './texts/Text'
import Title from './texts/Title'
import AppHeaderLarge from './views/AppHeaderLarge'
import ModularView from './views/ModularView'
import SafeAreaScrollView from './views/SafeAreaScrollView'

export {
  AppHeaderLarge,
  CheckBoxRow,
  ContactListItem,
  CredentialListItem,
  Label,
  Failure,
  Message,
  Pending,
  Success,
  ModularView,
  Button,
  NotificationCredentialListItem,
  NotificationProofListItem,
  QRScanner,
  SafeAreaScrollView,
  TextInput,
  Text,
  Title,
}
