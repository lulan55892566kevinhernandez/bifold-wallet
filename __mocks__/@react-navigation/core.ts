const navigation = {
  navigate: jest.fn(),
  goBack: jest.fn(),
  pop: jest.fn(),
}

const useNavigation = () => {
  return navigation
}

export { useNavigation }
