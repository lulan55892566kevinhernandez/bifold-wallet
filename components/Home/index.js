import React, {useState, useEffect, useContext} from 'react'

import {
  SafeAreaView,
  ScrollView,
  StatusBar,
  Text,
  TouchableOpacity,
  View,
} from 'react-native'

import {useHistory, useLocation} from 'react-router-native'

import AppHeaderLarge from '../AppHeaderLarge/index.js'
import BackButton from '../BackButton/index.js'
import LoadingOverlay from '../LoadingOverlay/index.js'

import {ErrorsContext} from '../Errors/index.js'
import {NotificationsContext} from '../Notifications/index.js'

import AppStyles from '../../assets/styles'

function Home() {
  let history = useHistory()
  let location = useLocation()

  const errors = useContext(ErrorsContext)
  const notifications = useContext(NotificationsContext)

  return (
    <>
    <BackButton backExit={true} />
    <View style={AppStyles.viewFull}>
      <AppHeaderLarge />
      <View style={AppStyles.tab}>
        <Text style={[AppStyles.h2, AppStyles.textBlueDark]}>Home</Text>
        <TouchableOpacity
          style={[AppStyles.button, AppStyles.buttonGreen, {marginTop: 80}]}
          onPress={() => {
            /*errors.setVisible(true)
            errors.setText("Workflows not\nyet created")
            errors.setPath("/home")*/
            history.push('/workflow/connect')
          }}>
          <Text style={[AppStyles.h2, AppStyles.textWhite]}>Begin</Text>
        </TouchableOpacity>
      </View>
    </View>
    </>
  )
}

export default Home
