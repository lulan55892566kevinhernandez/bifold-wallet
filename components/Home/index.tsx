import React, { useContext } from 'react'

import { Text, TouchableOpacity, View } from 'react-native'

import { useHistory, useLocation } from 'react-router-native'

import AppHeaderLarge from '../AppHeaderLarge/index'
import BackButton from '../BackButton/index'

import { ErrorsContext } from '../Errors/index'
import { NotificationsContext } from '../Notifications/index'

import AppStyles from '../../assets/styles'

function Home() {
  const history = useHistory()
  const location = useLocation()

  const errors = useContext(ErrorsContext)
  const notifications = useContext(NotificationsContext)

  return (
    <>
      <BackButton backExit={true} />
      <View style={AppStyles.viewFull}>
        <View style={[AppStyles.tab, { top: 80 }]}>
          <AppHeaderLarge />
          <TouchableOpacity
            style={[AppStyles.button, AppStyles.backgroundPrimary, { marginTop: 10 }]}
            onPress={() => {
              /*errors.setVisible(true)
            errors.setText("Workflows not\nyet created")
            errors.setPath("/home")*/
              history.push('/workflow/connect')
            }}
          >
            <Text style={[AppStyles.h2, AppStyles.textWhite]}>Begin</Text>
          </TouchableOpacity>
        </View>
      </View>
    </>
  )
}

export default Home
