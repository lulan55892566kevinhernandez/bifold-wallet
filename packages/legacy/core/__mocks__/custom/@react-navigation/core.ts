const navigate = jest.fn()
const dispatch = jest.fn()
const replace = jest.fn()
const navigation = {
  __timestamp: process.hrtime(),
  navigate,
  replace,
  setOptions: jest.fn(),
  getParent: () => {return {
    navigate,
    dispatch,
    replace,
  }},
  getState: jest.fn(() => ({
    index: jest.fn(),
  })),
  goBack: jest.fn(),
  pop: jest.fn(),
  reset: jest.fn(),
  isFocused: () => true,
  dispatch,
}

const useNavigation = () => {
  return navigation
}

const useIsFocused = () => {
  return true
}

export { useNavigation, useIsFocused }
