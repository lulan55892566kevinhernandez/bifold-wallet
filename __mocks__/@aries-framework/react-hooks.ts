import { CredentialRecord, ProofRecord } from '@aries-framework/core'

const useCredentials = jest.fn().mockReturnValue({ credentials: [] } as any)
const useCredentialByState = jest.fn().mockReturnValue([] as CredentialRecord[])
const useProofByState = jest.fn().mockReturnValue([] as ProofRecord[])
const mockAgent = {
  acceptOffer: jest.fn(),
  declineOffer: jest.fn(),
}
const useAgent = () => ({
  agent: {
    credentials: mockAgent,
  },
})
const useCredentialById = jest.fn()

export { useAgent, useCredentialById, useCredentials, useCredentialByState, useProofByState }
