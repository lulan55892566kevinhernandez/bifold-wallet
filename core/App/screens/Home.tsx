import { StackScreenProps } from '@react-navigation/stack'
import React from 'react'
import { useTranslation } from 'react-i18next'
import { FlatList, StyleSheet, View, Text, Dimensions, TouchableOpacity } from 'react-native'
import { ScrollView } from 'react-native-gesture-handler'

import NotificationListItem, { NotificationType } from '../components/listItems/NotificationListItem'
import NoNewUpdates from '../components/misc/NoNewUpdates'
import { useConfiguration } from '../contexts/configuration'
import { useTheme } from '../contexts/theme'
import { HomeStackParams, Screens } from '../types/navigators'

const { width } = Dimensions.get('window')
const offset = 25
const offsetPadding = 5

type HomeProps = StackScreenProps<HomeStackParams, Screens.Home>

const Home: React.FC<HomeProps> = ({ navigation }) => {
  const { useCustomNotifications } = useConfiguration()
  const { notifications } = useCustomNotifications()
  const { t } = useTranslation()
  const { homeContentView: HomeContentView } = useConfiguration()
  // This syntax is required for the jest mocks to work
  // eslint-disable-next-line import/no-named-as-default-member
  const { HomeTheme } = useTheme()

  const styles = StyleSheet.create({
    container: {
      paddingHorizontal: offset,
    },
    rowContainer: {
      flexDirection: 'row',
      justifyContent: 'space-between',
      alignItems: 'center',
      flexWrap: 'wrap',
      paddingHorizontal: offset,
    },
    messageContainer: {
      alignItems: 'center',
      justifyContent: 'center',
      marginTop: 35,
      marginHorizontal: offset,
    },
    header: {
      marginTop: offset,
      marginBottom: 20,
    },
    linkContainer: {
      minHeight: HomeTheme.link.fontSize,
      marginTop: 10,
    },
    link: {
      ...HomeTheme.link,
    },
  })

  const DisplayListItemType = (item: any): Element => {
    let component: Element
    if (item.type === 'CredentialRecord') {
      let notificationType = NotificationType.CredentialOffer
      if (item.revocationNotification) {
        notificationType = NotificationType.Revocation
      }
      component = <NotificationListItem notificationType={notificationType} notification={item} />
    } else if (item.type === 'CustomNotification') {
      component = <NotificationListItem notificationType={NotificationType.Custom} notification={item} />
    } else {
      component = <NotificationListItem notificationType={NotificationType.ProofRequest} notification={item} />
    }
    return component
  }

  return (
    <>
      <ScrollView>
        <View style={styles.rowContainer}>
          <Text style={[HomeTheme.notificationsHeader, styles.header]}>
            {t('Home.Notifications')}
            {notifications?.length ? ` (${notifications.length})` : ''}
          </Text>
          {notifications?.length > 1 ? (
            <TouchableOpacity
              style={styles.linkContainer}
              activeOpacity={1}
              onPress={() => navigation.navigate(Screens.Notifications)}
            >
              <Text style={styles.link}>{t('Home.SeeAll')}</Text>
            </TouchableOpacity>
          ) : null}
        </View>
        <FlatList
          horizontal
          showsHorizontalScrollIndicator={false}
          scrollEnabled={notifications?.length > 0 ? true : false}
          snapToOffsets={[
            0,
            ...Array(notifications?.length)
              .fill(0)
              .map((n: number, i: number) => i * (width - 2 * (offset - offsetPadding)))
              .slice(1),
          ]}
          decelerationRate="fast"
          ListEmptyComponent={() => (
            <View style={{ marginHorizontal: offset, width: width - 2 * offset }}>
              <NoNewUpdates />
              <View style={[styles.messageContainer]}>
                <Text adjustsFontSizeToFit style={[HomeTheme.welcomeHeader, { marginTop: offset, marginBottom: 20 }]}>
                  {t('Home.Welcome')}
                </Text>
              </View>
            </View>
          )}
          data={notifications}
          keyExtractor={(item) => item.id}
          renderItem={({ item, index }) => (
            <View
              accessible={true}
              style={{
                width: width - 2 * offset,
                marginLeft: !index ? offset : offsetPadding,
                marginRight: index === notifications?.length - 1 ? offset : offsetPadding,
              }}
            >
              {DisplayListItemType(item)}
            </View>
          )}
        />
        <HomeContentView />
      </ScrollView>
    </>
  )
}

export default Home
