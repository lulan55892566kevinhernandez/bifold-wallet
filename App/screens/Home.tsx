import { CredentialState } from '@aries-framework/core'
import { useCredentialByState } from '@aries-framework/react-hooks'
import { StackScreenProps } from '@react-navigation/stack'
import React from 'react'
import { useTranslation } from 'react-i18next'
import { FlatList, StyleSheet, View, Text, Dimensions, TouchableOpacity } from 'react-native'

import { NotificationType } from '../components/listItems/NotificationListItem'
import { useNotifications } from '../hooks/notifcations'
import { ColorPallet, TextTheme } from '../theme'
import { HomeStackParams, Screens } from '../types/navigators'

import { InfoTextBox, NotificationListItem } from 'components'

const { width } = Dimensions.get('window')

const offset = 25
const offsetPadding = 5

type HomeProps = StackScreenProps<HomeStackParams, Screens.Home>

const styles = StyleSheet.create({
  container: {
    paddingHorizontal: offset,
  },
  rowContainer: {
    flexDirection: 'row',
    justifyContent: 'space-between',
    alignItems: 'center',
    paddingHorizontal: offset,
  },
  messageContainer: {
    alignItems: 'center',
    justifyContent: 'center',
    marginTop: 35,
    marginHorizontal: 60,
  },
  header: {
    marginTop: offset,
    marginBottom: 20,
  },
  linkContainer: {
    minHeight: TextTheme.normal.fontSize,
    marginTop: 10,
  },
  link: {
    ...TextTheme.normal,
    color: ColorPallet.brand.link,
  },
})

const Home: React.FC<HomeProps> = ({ navigation }) => {
  const credentials = [
    ...useCredentialByState(CredentialState.CredentialReceived),
    ...useCredentialByState(CredentialState.Done),
  ]

  const { notifications } = useNotifications()
  const { t } = useTranslation()

  const emptyListComponent = () => (
    <View style={{ marginHorizontal: offset, width: width - 2 * offset }}>
      <InfoTextBox>
        <Text style={[TextTheme.normal]}>{t('Home.NoNewUpdates')}</Text>
      </InfoTextBox>
    </View>
  )

  const displayMessage = (credentialCount: number) => {
    if (typeof credentialCount === 'undefined' && credentialCount >= 0) {
      throw new Error('Credential count cannot be undefined')
    }

    let credentialMsg

    if (credentialCount === 1) {
      credentialMsg = (
        <Text>
          {t('Home.YouHave')} <Text style={{ fontWeight: 'bold' }}>{credentialCount}</Text> {t('Home.Credential')}{' '}
          {t('Home.InYourWallet')}
        </Text>
      )
    } else if (credentialCount > 1) {
      credentialMsg = (
        <Text>
          {t('Home.YouHave')} <Text style={{ fontWeight: 'bold' }}>{credentialCount}</Text> {t('Home.Credentials')}{' '}
          {t('Home.InYourWallet')}
        </Text>
      )
    } else {
      credentialMsg = t('Home.NoCredentials')
    }

    return (
      <View style={[styles.messageContainer]}>
        {credentialCount === 0 ? (
          <Text style={[TextTheme.headingOne, { marginTop: offset, marginBottom: 35 }]}>{t('Home.Welcome')}</Text>
        ) : null}
        <Text style={[TextTheme.normal, { marginTop: offset, textAlign: 'center' }]}>{credentialMsg}</Text>
      </View>
    )
  }

  return (
    <View>
      <View style={styles.rowContainer}>
        <Text style={[TextTheme.headingThree, styles.header]}>
          {t('Home.Notifications')}
          {notifications?.length ? ` (${notifications.length})` : ''}
        </Text>
        {notifications?.length > 1 ? (
          <TouchableOpacity
            style={styles.linkContainer}
            activeOpacity={1}
            onPress={() => navigation.navigate(Screens.Notifications)}
          >
            <Text style={styles.link}>{t('Home.SeeAll')}</Text>
          </TouchableOpacity>
        ) : null}
      </View>
      <FlatList
        horizontal
        showsHorizontalScrollIndicator={false}
        scrollEnabled={notifications?.length > 0 ? true : false}
        snapToOffsets={[
          0,
          ...Array(notifications?.length)
            .fill(0)
            .map((n: number, i: number) => i * (width - 2 * (offset - offsetPadding)))
            .slice(1),
        ]}
        decelerationRate="fast"
        ListEmptyComponent={emptyListComponent()}
        data={notifications}
        keyExtractor={(item) => item.id}
        renderItem={({ item, index }) => (
          <View
            style={{
              width: width - 2 * offset,
              marginLeft: !index ? offset : offsetPadding,
              marginRight: index === notifications?.length - 1 ? offset : offsetPadding,
            }}
          >
            {item.type === 'CredentialRecord' ? (
              <NotificationListItem notificationType={NotificationType.CredentialOffer} notification={item} />
            ) : (
              <NotificationListItem notificationType={NotificationType.ProofRequest} notification={item} />
            )}
          </View>
        )}
      />
      <View style={styles.container}>{displayMessage(credentials.length)}</View>
    </View>
  )
}

export default Home
