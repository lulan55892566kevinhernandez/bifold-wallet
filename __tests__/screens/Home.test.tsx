import { CredentialRecord, CredentialState, ProofRecord } from '@aries-framework/core'
import { useCredentialByState, useProofByState } from '@aries-framework/react-hooks'
import { useNavigation } from '@react-navigation/core'
import { fireEvent, render } from '@testing-library/react-native'
import React from 'react'
import { FlatList, TouchableOpacity } from 'react-native'
import { create } from 'react-test-renderer'

import Home from '../../App/screens/Home'

import { InfoTextBox, NotificationListItem } from 'components'

describe('displays a home screen', () => {
  it('renders correctly', () => {
    const tree = create(<Home />).toJSON()

    expect(tree).toMatchSnapshot()
  })

  // TODO:(jl) Good enough to be captured by the snapshot?
  it('defaults to no notifications', () => {
    const tree = create(<Home />)
    const root = tree.root
    const results = root.findByType(InfoTextBox)

    expect(results.props.children).toBe('Home.NoNewUpdates')
  })
})

describe('with a notifications module, when an issuer sends a credential offer', () => {
  const testCredentialRecords: CredentialRecord[] = [
    new CredentialRecord({
      threadId: '1',
      state: CredentialState.OfferReceived,
    }),
  ]
  const testProofRecords: ProofRecord[] = []

  beforeEach(() => {
    jest.clearAllMocks()
    // @ts-ignore
    useCredentialByState.mockReturnValue(testCredentialRecords)
    // @ts-ignore
    useProofByState.mockReturnValue(testProofRecords)
  })

  /**
   * Scenario: Holder receives a credential offer
   * Given the holder has a connection with an issuer
   * When the issuer sends a credential offer
   * Then the credential offer will arrive in the form of a notification in the home screen
   */
  it('notifications are displayed', () => {
    const tree = create(<Home />)
    const root = tree.root
    const flatListInstance = root.findByType(FlatList)

    expect(flatListInstance.findAllByType(NotificationListItem)).toHaveLength(1)
  })

  /**
   * Scenario: Holder selects a credential offer
   * Given the holder has received a credential offer from a connected issuer
   * When the holder selects on the credential offer
   * When the holder is taken to the credential offer screen/flow
   */
  it('touch notification triggers navigation correctly', async () => {
    const tree = create(<Home />)
    const root = tree.root
    const touchable = root.findByType(NotificationListItem).findByType(TouchableOpacity)
    const navigation = useNavigation()

    touchable.props.onPress()

    expect(navigation.navigate).toHaveBeenCalledTimes(1)
    expect(navigation.navigate).toHaveBeenCalledWith('Credential Offer', { credentialId: testCredentialRecords[0].id })
  })
})
