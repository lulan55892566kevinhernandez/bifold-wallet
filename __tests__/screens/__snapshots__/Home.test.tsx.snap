// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`displays a home screen renders correctly 1`] = `
<View
  style={
    Object {
      "paddingHorizontal": 25,
    }
  }
>
  <Text
    style={
      Array [
        Object {
          "color": "#FFFFFF",
          "fontSize": 26,
          "fontWeight": "bold",
        },
        Object {
          "marginBottom": 20,
          "marginTop": 25,
        },
      ]
    }
  >
    Home.Notifications
  </Text>
  <RCTScrollView
    ListEmptyComponent={
      <InfoTextBox>
        Home.NoNewUpdates
      </InfoTextBox>
    }
    data={Array []}
    getItem={[Function]}
    getItemCount={[Function]}
    horizontal={true}
    keyExtractor={[Function]}
    onContentSizeChange={[Function]}
    onLayout={[Function]}
    onMomentumScrollBegin={[Function]}
    onMomentumScrollEnd={[Function]}
    onScroll={[Function]}
    onScrollBeginDrag={[Function]}
    onScrollEndDrag={[Function]}
    removeClippedSubviews={false}
    renderItem={[Function]}
    scrollEnabled={false}
    scrollEventThrottle={50}
    showsHorizontalScrollIndicator={false}
    stickyHeaderIndices={Array []}
    viewabilityConfigCallbackPairs={Array []}
  >
    <View>
      <View
        style={
          Array [
            Object {
              "backgroundColor": "rgba(53, 130, 63, 0.7)",
              "borderColor": undefined,
              "borderRadius": 5,
              "borderWidth": 1,
              "flexDirection": "row",
              "padding": 10,
            },
          ]
        }
      >
        <View
          style={
            Array [
              Object {
                "marginRight": 10,
              },
            ]
          }
        >
          <Text
            allowFontScaling={false}
            style={
              Array [
                Object {
                  "color": "#FFFFFF",
                  "fontSize": 30,
                },
                undefined,
                Object {
                  "fontFamily": "Material Icons",
                  "fontStyle": "normal",
                  "fontWeight": "normal",
                },
                Object {},
              ]
            }
          >
            î¢Ž
          </Text>
        </View>
        <Text
          style={
            Array [
              Object {
                "color": "#FFFFFF",
                "flexShrink": 1,
                "fontSize": 18,
                "fontWeight": "normal",
              },
              Object {
                "fontWeight": "bold",
              },
            ]
          }
        >
          Home.NoNewUpdates
        </Text>
      </View>
    </View>
  </RCTScrollView>
  <View
    style={
      Array [
        Object {
          "alignItems": "center",
          "justifyContent": "center",
          "marginHorizontal": 60,
          "marginTop": 35,
        },
      ]
    }
  >
    <Text
      style={
        Array [
          Object {
            "color": "#FFFFFF",
            "fontSize": 38,
            "fontWeight": "bold",
          },
        ]
      }
    >
      Home.Welcome
    </Text>
    <Text
      style={
        Array [
          Object {
            "color": "#FFFFFF",
            "fontSize": 18,
            "fontWeight": "normal",
          },
          Object {
            "marginTop": 25,
            "textAlign": "center",
          },
        ]
      }
    >
      Home.NoCredentials
    </Text>
  </View>
</View>
`;
